crates/matrix-sdk/src/widget/machine/from_widget.rs

```diff
@@ -41,28 +41,80 @@ pub(super) enum FromWidgetRequest {
     DelayedEventUpdate(UpdateDelayedEventRequest),
 }
 
+/// The full response a client sends to a from widget request
+/// in case of an error.
 #[derive(Serialize)]
 pub(super) struct FromWidgetErrorResponse {
     error: FromWidgetError,
 }
 
 impl FromWidgetErrorResponse {
-    pub(super) fn new(e: impl fmt::Display) -> Self {
-        Self { error: FromWidgetError { message: e.to_string() } }
+    /// Create a error response to send to the widget from an http error.
+    pub(crate) fn from_http_error(error: &HttpError) -> Self {
+        Self {
+            error: FromWidgetError {
+                message: error.to_string(),
+                matrix_api_error: error.as_client_api_error().and_then(
+                    |api_error| match &api_error.body {
+                        ErrorBody::Standard { kind, message } => Some(FromWidgetMatrixErrorBody {
+                            http_status: api_error.status_code.as_u16().into(),
+                            response: json!({"errcode": kind.to_string(), "error": message }),
+                        }),
+                        _ => None,
+                    },
+                ),
+            },
+        }
+    }
+
+    /// Create a error response to send to the widget from a matrix sdk error.
+    pub(crate) fn from_error(error: &Error) -> Self {
+        match &error {
+            Error::Http(e) => FromWidgetErrorResponse::from_http_error(e),
+            Error::UnknownError(e) => FromWidgetErrorResponse::from_string(e.to_string()),
+            _ => FromWidgetErrorResponse::from_string(error.to_string()),
+        }
+    }
+
+    /// Create a error response to send to the widget from a string.
+    pub(crate) fn from_string<S: Into<String>>(error: S) -> Self {
+        Self { error: FromWidgetError { message: error.into(), matrix_api_error: None } }
     }
 }
 
+/// The serializable section of an error response send by the client as a
+/// response to a from widget request.
```

Comment:
Unless this is referring to a specific type:

```suggestion
/// response to a request from the widget.
```

If this is referring to a specific type, could it be the type name with a link, e.g. ```[`TheTypeName`]```?